README.md
# clota: aka iota.lib.clj

An [IOTA](https://iota.org) Client Library Implemented in Clojure

This library aims to be an idiomatic, feature-complete, client implementation of the full IOTA API Command Set.

Show your support for this library and this emerging protocol by giving this repo a **star**

## Getting Started

Get Clota from clojars:

[![Clojars Project](https://img.shields.io/clojars/v/clota.svg)](https://clojars.org/clota)

or download and install directly:

```bash
git clone https://github.com/tuddman/iota.lib.clj
cd iota.lib.clj
lein clean && lein install
```

## Usage

```clojure
;; Include Clota Library in your project

> (require '[clota.client :as iota])

;; Commands - from a REPL

> (def node-uri "http://0.0.0.0:14625")

;; OPTIONAL - if you have REMOTE_AUTH set for the node. If so, you will need to send the token along with every command.

> (def token (iota/->base64 "user:password")

> (iota/get-node-info node-uri token)

> (iota/get-neighbors node-uri token)

> (iota/add-neighbors node-uri ["udp://good.neighbor:14265" ...] token)

> (iota/remove-neighbors node-uri ["udp://bad.neighbor:14265" ...] token)

> (iota/get-tips node-uri token)

> (iota/find-transactions node-uri {:addresses ["SOMEIOTAHASHADDRESS..." ...]} token)

> (iota/get-trytes node-uri ["SOMEIOTATRYTES...999"] token)

> (iota/get-inclusion-states node-uri {:transactions ["..." ...] :tips ["..." ...]} token)

> (iota/get-balances node-uri {:addresses ["..." ...] :threshold 100} token)

> (iota/get-transactions-to-approve node-uri {:depth 27} token)

> (iota/attach-to-tangle node-uri {:trunkTransaction "..." :branchTransaction "..." :min-weight-magnitude 18 :trytes ["..." ...]} token)

> (iota/interrupt-attaching-to-tangle node-uri token)

> (iota/broadcast-transactions node-uri ["SOMETXHASH..." ...] token)

> (iota/store-transactions node-uri ["SOMETXHASH..." ...] token)

```

[read the API Docs](https://iota.readme.io)

API Command | Clota Function | Status | Test Coverage
--- | --- | --- | ---
[getNodeInfo](https://iota.readme.io/docs/getnodeinfo) | `(get-node-info ...)` | Implemented | Complete
[getNeighbors](https://iota.readme.io/docs/getneighborsactivity) | `(get-neighbors ...)` | Implemented | Complete
[removeNeighbors](https://iota.readme.io/docs/getnodeinfo) | `(remove-neighbors ...)` | Implemented | Complete
[getTips](https://iota.readme.io/docs/gettips) | `(get-tips ...)` | Implemented | Complete
[findTransactions](https://iota.readme.io/docs/findtransactions) | `(find-transactions ...)` | Implemented | None
[getTrytes](https://iota.readme.io/docs/gettrytes) | `(get-trytes ...)` | Implemented | None
[getInclusionStates](https://iota.readme.io/docs/getinclusionstates) | `(get-inclusion-states ...)` | Implemented | None
[getBalances](https://iota.readme.io/docs/getbalances) | `(get-balances ...)` | Implemented | None
[getTransactionsToApprove](https://iota.readme.io/docs/gettransactionstoapprove) | `(get-transactions-to-approve ...)` | Implemented | None
[attachToTangle](https://iota.readme.io/docs/attachtotangle) | `(attach-to-tangle ...)` | Implemented | None
[interruptAttachingToTangle](https://iota.readme.io/docs/interruptattachingtotangle) | `(interrupt-attaching-to-tangle ...)` | Implemented | None
[broadcastTransactions](https://iota.readme.io/docs/broadcasttransactions) | `(broadcast-transactions ...)` | Implemented | None
[storeTransactions](https://iota.readme.io/docs/storetransactions) | `(store-transactions ...)` | Implemented | None

for more, look in `src/clota/client.clj`

### Generate a Secure Random Seed 

```clojure
;; PLEASE BE ADVISED : DO NOT SHARE YOUR PRIVATE SEED WITH ANYONE. EVER!
;; To Generate a Secure Random Seed:

> (clota.crypto/generate-seed)

 =>  "9KHKESVOHBOMHCL9ADINALYVHSBKRHITKMOJJJHFKOHWBZDFQDVLKQMTVZTUTWHIWXISBRIFXPZEEZTUU" 

;; Or, if that's not good enough for 'ya, 
;; and you want to do some more randomized matrix math to generate your seed:

> (clota.crypto/generate-paranoid-seed)

 =>  "IKAGOIVHXAR9XRJMC9LIXHMFAXYAAXKSAXJE99ODVRMISXWNJIIGGVCQIKLVHMRQRQKESJDESRSQOUZTA"
```

## Testing

Continuous Integration Tests are provided by [Travis CI.](https://travis-ci.com/)

[![Build Status](https://travis-ci.org/tuddman/iota.lib.clj.svg?branch=master)](https://travis-ci.org/tuddman/iota.lib.clj)


To run the tests yourself:

```bash
$ lein test
```


Pull Requests Welcome.


---

Copyright © 2017 - now tuddman

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log

## 0.1.2 
### Added
- generate seed QR codes

## [0.1.1] - 2017-09-15
### Added
- Authenticated (token) Requests

### Fixed
- Bug fixes with optional arity parameters 

## 0.1.0 - 2017-08-23
- Initial Release
- 14 IOTA API commands supported
- Secure Seed Generator
- Paranoid Secure Seed Generator

[0.1.1]: https://clojars.org/clota 


test/clota/core_test.clj
(ns clota.client-test
  (:require [clojure.test :refer :all]
            [clota.client :refer :all]
            [cheshire.core :as json]
            [clj-http.client :as client]))

(deftest test-get-node-info
  (is (= (get-node-info "http://localhost:14700") {})))


test/clota/client_test.clj
(ns clota.client-test
  (:require [clojure.test :refer :all]
            [clota.client :refer :all]
            [cheshire.core :as json]
            [clj-http.client :as client]
            [environ.core :refer [env]]))

(def localhost "http://localhost:14700")
(def ci-host-node (env :iota-test-node))
(def ci-host-neighbor (env :iota-test-neighbor))

(def test-node ci-host-node)

(deftest test-get-node-info
  (let [host test-node]
  (testing "does get-node-info return data"
    (is (= "IRI" (-> (get-node-info host) :appName))))))

(deftest test-get-neighbors
  (let [host test-node]
  (testing "does get-node-info return data"
    (is (<= 0 (-> (get-neighbors host) :duration)))
    #_(is (not-empty (-> (get-neighbors host) :neighbors))))))

(deftest test-add-neighbors
  (let [host test-node
        neighbor ci-host-neighbor
        added-neighbor-resp (add-neighbors host [neighbor])
        neighbors (map :address (:neighbors (get-neighbors test-node)))]
  (testing "does a neighbor get added"
    #_(is (= 1 (:addedNeighbors added-neighbor-resp)))
    #_(is (= true (not (empty (some #{neighbor} neighbors)))) ))))

(deftest test-remove-neighbors
  (let [host test-node
        neighbor ci-host-neighbor
        remove-neighbor-resp (remove-neighbors host [neighbor])
        neighbors (map :address (:neighbors (get-neighbors test-node)))]
  (testing "does a neighbor get removed"
    #_(is (= 1 (:removedNeighbors remove-neighbor-resp)))
    #_(is (= nil (some #{neighbor} neighbors)) ))))

(deftest test-get-tips
  (let [host test-node]
  (testing "does get-tips return a vector of tip hashes"
    (is (not-empty (-> (get-tips host) :hashes))))))   

(deftest test-find-transactions
  (let [host test-node]
  (testing ""
    (is (= true (or true "test not yet implmented"))))))

(deftest test-get-trytes
  (let [host test-node]
  (testing ""
    (is (= true (or true "test not yet implmented"))))))

(deftest test-get-inclusion-states
  (let [host test-node]
  (testing ""
    (is (= true (or true "test not yet implmented"))))))

(deftest test-get-balances
  (let [host test-node]
  (testing ""
    (is (= true (or true "test not yet implmented"))))))

(deftest test-get-transactions-to-approve
  (let [host test-node]
  (testing ""
    (is (= true (or true "test not yet implmented"))))))

(deftest test-attach-to-tangle
  (let [host test-node]
  (testing ""
    (is (= true (or true "test not yet implmented"))))))

(deftest test-interrupt-attaching-to-tangle
  (let [host test-node]
  (testing ""
    (is (= true (or true "test not yet implmented"))))))

(deftest test-broadcast-transactions
  (let [host test-node]
  (testing ""
    (is (= true (or true "test not yet implmented"))))))

(deftest test-store-transactions
  (let [host test-node]
  (testing ""
    (is (= true (or true "test not yet implmented"))))))


test/clota/crypto_test.clj
(ns clota.crypto-test
  (:require [clojure.test :refer :all]
            [clota.crypto :refer :all]))

(deftest test-generate-seed
  (let [seed (generate-seed)]
  (testing "does seed generation work"
    (is (= 81 (count seed)))
    (is (identical? seed (re-matches #"[9A-Z]{0,81}" seed))))))


(deftest test-generate-paranoid-seed
  (let [seed (generate-paranoid-seed)]
  (testing "does seed generation work"
    (is (= 81 (count seed)))
    (is (identical? seed (re-matches #"[9A-Z]{0,81}" seed))))))






src/clota/crypto.clj
(ns clota.crypto
  (:import java.security.SecureRandom))
  
(defn trytes
  [size]
  (let [tryte-alphabet  "9ABCDEFGHIJKLMNOPQRSTUVWXYZ"]
    (apply str (take size (repeatedly #(.charAt tryte-alphabet (.nextInt (SecureRandom.) 27)))))))

(defn generate-seed []
  (trytes 81))

(defn generate-paranoid-seed
  "this fn generates a seed by:
    - Forms an 81x81 matrix of seeds. Each of which could be used as perfectly secure seed. But why stop there?
    - Securely & randomly selects an element from each row. Result is also a perfectly secure seed. 
   Some would say this is overkill to get to the same-level-of-security seed as (generate-seed).
   Those people aren't paranoid and this function is not for them. 
   This function is for you, my paranoid friend."
  []
  (let [rand-selector-fn (fn [seed] (nth seed (.nextInt (SecureRandom.) (count seed))))
        m (take 81 (repeatedly generate-seed))]
    (->> m
      (map rand-selector-fn)
      (apply str))))


src/clota/client.clj
(ns clota.client
  (:require  
    [base64-clj.core :as base64]
    [clojure.spec.alpha :as s]
    [cheshire.core :as json]
    [clj-http.client :as client]))


(defn ->base64
  [s]
  (base64/encode s))

(def fields
  #{"address" "addresses" "approvees" "bundles" "hash" "message" "minWeightMagnitude" "securityLevel" "seed" "tags" "transaction" "trytes" "uris"})

(def iota-localhost "http://localhost:14625")

;; from https://iota.readme.io/docs

(defn get-iota-version [] (json/encode {:version "1.2.0"}))

(defn- build-iota-req
  "params 
    - command { String }
    - params { Map } Optional "
  [{:keys [host command token] :or {host iota-localhost}} & [params]]
  (let [body (json/encode (merge {:command command} (or params {})))
        req (try (client/post (str host)
                              {:body body
                               :headers {"Authorization" (str "Basic " token)}
                               :content-type :json })
                 (catch Exception e (ex-data e)))]
    (-> req
        :body
        (json/decode true))))

(defn get-node-info 
  "@param { String } host" 
  [host & [token]]          
  (build-iota-req {:host host
                   :token token
                   :command "getNodeInfo"}))

(defn get-neighbors 
  "@param { String } host" 
  [host & [token]] 
  (build-iota-req {:host host
                   :token token
                   :command "getNeighbors"}))

(defn add-neighbors 
  "@param { String } host 
   @param { Vector of Strings } uris"
 [host [uris] & [token]] 
  (build-iota-req {:host host
                   :token token
                   :command "addNeighbors"}
                  {:uris [uris]}))

(defn remove-neighbors
  "@param { String } host 
   @param { Vector of Strings } uris" 
  [host [uris] & [token]] 
  (build-iota-req {:host host
                   :token token
                   :command "removeNeighbors"}
                  {:uris [uris]}))

(defn get-tips 
  "@param { String } host" 
  [host & [token]] 
  (build-iota-req {:host host
                   :token token
                   :command "getTips"}))

(defn find-transactions 
  "find-transactions can take as inputs a map of one or more of:
    - bundles [...] : List of bundle hashes. The hashes needs to be extended to 81 trytes by padding the hash with 9's.
    - addresses [...]  : List of addresses.
    - tags [...] : List of tags. Has to be 27 trytes. 
    - approvees [...] : List of approvee transaction hashes.  
  
   Note: Including more than one input type returns the intersection of these values.
  
   e.g. (find-transaction host {:addresses [AA9... AB9...]}d
   @param { String } host 
   @param { Map } :bundles :addresses :tags :approvees" 
  [host inputs & [token]] 
  (build-iota-req {:host host
                   :token token
                   :command "findTransactions"}
                  inputs))
 
(defn get-trytes 
  "@param { String } host
   @param { Vector of Strings } trytes : List of transaction hashes of which you want to get trytes from."
  [host [trytes] & [token]] 
  (build-iota-req {:host host
                   :token token
                   :command "getTrytes"}
                  {:hashes [trytes]}))

(defn get-inclusion-states 
  "@param { String } host" 
  [host {:keys [transactions tips] :as params} & [token]] 
  (build-iota-req {:host host
                   :token token
                   :command "getInclusionStates"}
                  params))

(defn get-balances 
  "@param { String } host" 
  [host {:keys [addresses threshold] :as params} & [token]] 
  (build-iota-req {:host host
                   :token token
                   :command "getBalances"}
                  params))

(defn get-transactions-to-approve 
  "@param { String } host
   @param { Map } :depth" 
  [host {:keys [depth] :as params} & [token]] 
  (build-iota-req {:host host
                   :token token
                   :command "getTransactionsToApprove"}
                  params))

(defn attach-to-tangle 
  "@param { String } host 
   @param { Map } :trunk-transaction :branch-transaction :min-weight-magnitude :trytes)" 
  [host {:keys [trunk-transaction branch-transaction min-weight-magnitude trytes] :as params} & [token]] 
  (build-iota-req {:host host
                   :token token
                   :command "attachToTangle"}
                  {:trunkTransaction trunk-transaction
                   :branchTransaction branch-transaction
                   :minWeightMagnitude min-weight-magnitude
                   :trytes trytes}))

(defn interrupt-attaching-to-tangle  
  "@param { String } host" 
  [host & [token]] 
  (build-iota-req {:host host
                   :token token
                   :command "interruptAttachingToTangle"}))

(defn broadcast-transactions
  "@param { String } host 
   @param { Vector of Strings } transactions : List of raw data of transactions to be rebroadcast."
  [host [transactions] & [token]] 
  (build-iota-req {:host host
                   :token token
                   :command "broadcastTransactions"}
                  {:trytes [transactions]}))

(defn store-transactions  
  "@param { Vector } transactions : List of raw data of transactions to store locally."
  [host [transactions] & [token]] 
  (build-iota-req {:host host
                   :token token
                   :command "storeTransactions"}
                  {:trytes [transactions]}))

